### PR TITLE
121 dodal integration

### LINKED ISSUE #121: Dodal integration
Should be possible to make devices from a dodal beamline module.
https://github.com/DiamondLightSource/blueapi/issues/120

### PR DESCRIPTION
**Pull request by abbiemery**
Fixes #121 .

Needs to have [this](https://github.com/DiamondLightSource/dodal/pull/44) dodal pr to be merged to work, otherwise the devices that use other devices in the example start up will not work.

It also fails on the 3.8 tests, so maybe we remove the support for that first.

### CODE REVIEW

**Comment by callumforrester** [comment]
The CI should be fixed if you rebase on `main`

File: src/blueapi/config.py
Comment (callumforrester) [inline src/blueapi/config.py:26]: Shouldn't this be in `EnvironmentConfig`?

**Review by callumforrester — CHANGES_REQUESTED** [review]
Looking good! Only a few small items.

File: src/blueapi/config.py
Comment (callumforrester) [reply src/blueapi/config.py:26]: Nevermind, I can't read!

File: src/blueapi/startup/example.py
Comment (tpoliaw) [inline src/blueapi/startup/example.py:6]: Is the `name` field always required and is it always going to be the same as the function name?

File: src/blueapi/config.py
Comment (abbiemery) [reply src/blueapi/config.py:7]: I planned to do this but forgot, oops.

File: src/blueapi/core/context.py
Comment (abbiemery) [reply src/blueapi/core/context.py:10]: I did add it but accidentially removed it in my last commit, well spotted. 

Also this was discussed briefly at the time and the code was already there for us to use so @callumforrester suggested we use it. We then rewrote all of that code anyway so it became slightly moot. I would be in favour of having our own util for it, however it then is the same code and we risk incompatibility with dodal if they change stuff up on their end and our `make_all_devices` doesn't keep up with those.

Happy to do it either way and will listen to any compelling opinion.

File: src/blueapi/startup/example.py
Comment (abbiemery) [reply src/blueapi/startup/example.py:6]: No it is not required as an argument, as long as the device return has one. I just copied the dodal format. It also doesn't matter what the function is called. The function name is used to determine if others depend on it and instantiate the devices in order, but the resulting dictionary is `device.name: device`.

**Comment by abbiemery** [comment]
@callumforrester I made the changes you requested. I also added a better test case for the context which will cause the CI to fail untill the `make_all_devices` changes are implemented. Either by dodal or if we choose to migrate the method to the blueapi utils.

File: src/blueapi/startup/example.py
Comment (callumforrester) [reply src/blueapi/startup/example.py:6]: It's not always going to be the same as the function name: https://nsls-ii.github.io/ophyd/architecture.html#names

File: src/blueapi/core/context.py
Comment (callumforrester) [inline src/blueapi/core/context.py:70]: We want to be doing a different action based on the value of `source.type`, so something like:
```python
if source.type == "plans":
    self.with_plan_module(mod)
elif source.type == "dodal":
    self.with_device_module(mod)
else:
    ...
```

File: src/blueapi/core/context.py
Comment (abbiemery) [reply src/blueapi/core/context.py:9]: I spoke to mx and agree we should leave the util in dodal.

**Comment by abbiemery** [comment]
@callumforrester Done the last things you pointed out and rebased. The tests will still fail until the dodal updates are merged. I'm just waiting on a review for that.

File: src/blueapi/config.py
Comment (callumforrester) [inline src/blueapi/config.py:16]: Very minor, but enums count as constants under PEP 8 so names should in SCREAMING_SNAKE_CASE

```suggestion
    PLAN_FUNCTIONS = "planFunctions"
    DEVICE_FUNCTIONS = "deviceFunctions"
    DODAL = "dodal"
```

**Review by callumforrester — APPROVED** [review]
Provisionally approving pending dodal PR merge. Only minor comments now. Some docstrings on new classes/methods would also be good if you have time.

**Pull request merged**
